### DISCUSSION TITLE
Feedback

### DISCUSSION THREAD

**Discussion by ggerganov** [ggerganov/ggwave#4] [General]
Post any kind of feedback about `ggwave` or `Waver` - ideas, issues, suggestions, things to improve, etc.

**Comment by astrickash** [comment]
It is a great repository to work with and I am very much excited about future of this technology. I am trying to convert the same thing in python but I didn't understand how you captured the received audio frequency order, I mean when you capture audio between sound markers then how did you find the order of frequencies. I read all the description you have provided on the main page but didn't get this thing. Hope you can explain that a little too.

**Comment by ggerganov** [comment]
Thanks for the feedback!

Can you be more specific about:

> how did you find the order of frequencies

I am not sure I understand which order you are referring to.

Here is some more info, hoping to make things a bit more clear:
Imagine the full spectrum from 0Hz to 24kHz. We are sampling at 48kHz and doing FFT with 1024 bins.This means each bin has width <tt>dF = SampleRate/nBins = 46,875 Hz</tt>. For the audible protocols, we choose a starting frequency <tt>F0 = 1875.000 Hz</tt> - this is the 40th bin. We use 96 bins to represent 3 bytes of data - so bins <tt>[40, 136)</tt> contain information about 3 bytes.

How is the data encoded into sound?
First, consider 4 bits of data - they can represent <tt>2<sup>4</sup> = 16</tt> distinct values. Let's say they currently represent the value <tt>0 <= k < 16</tt>. In order to encode these 4 bits (i.e. the value <tt>k</tt>) into sound, we add an amplitude with frequency corresponding to the <tt>k</tt>-th bin. So for 4 bits of data, we need to use 16 bins from the spectrum.

Since we use 96 bins in total, we can encode 96/16 = 6 chunks of 4 bits which is exactly 3 bytes. We now split our payload that we want to transmit in chunks of 3 bytes and sequentially add the necessary amplitudes and transmit. Hope that makes sense.

While writing this, I realized I might have overlooked the way the bins are centered. I am assuming that bin <tt>i</tt> is centered at frequency <tt>i\*dF</tt>, but probably it might be actually centered at <tt>i\*dF + 0.5\*dF</tt>. Not sure if this is correct or if it has any impact on the performance of the method. Will see ..

**Reply by astrickash** [reply]
Thanks for your reply and this explanation. 
So basically what I understand is if we need to send `abc` then we need to send `0110 0001 0110 0010 0110 0011` which is actually 24 bit data or 6 chunks of 4bit data and because you are using different frequency range for sending different chunks you don't actually need to write the code for which character comes first in the audio. 
Also, yes the data is not centered at `i*df` , it is actually shifted towards positive x-axis but I didn't find any impact of this yet.

**Reply by ggerganov** [reply]
I looked a bit more into what frequencies the FFT bins really represent and it seems the original assumption is true - bin <tt>i</tt> is "centered" at frequency <tt>i*dF</tt>. Here is one article discussing this: [link](http://www.tedknowlton.com/resume/FFT_Bin_Interp.html#:~:text=*Bin%20center%20frequency%20is%20the,512%20%3D%209.766%20Hz%20%3D%20fbin.)

I also tested that if I generate a sine wave sound with frequency of exactly <tt>i*dF</tt> I get signal only in the <tt>i</tt>th bin of the FFT.

**Comment by CasualArclamp** [comment]
have you though of using Direct Sequence Spread spectrum like the mode ROS? https://rosmodem.wordpress.com/ its at the very bottom

**Reply by ggerganov** [reply]
Hi I wasn't aware of that technique - I guess I could try spreading the tones randomly and see if that helps. I guess I can also spread the ECC bytes - currently, they are at the end of the message, but I could apply the same technique and spread the ECC bytes in-between the message.

**Reply by ggerganov** [reply]
I recently implemented a similar technique in the [Waver](https://github.com/ggerganov/ggwave/tree/master/examples/waver) application:

![image](https://user-images.githubusercontent.com/1991296/155838848-d21f4605-9024-4be0-ac34-6985abb23749.png)

It works by xoring the payload with a magic sequence of bytes, causing the frequencies to be spread more uniformly within the frequency spectrum.

The implementation is here if anyone is interested: https://github.com/ggerganov/ggwave/commit/75e3fde357b4dff71c49ce91da94ca4d4edf5118

**Comment by AngryApples** [comment]
Fantastic library, @ggerganov. Kudos!

Is there a utility yet to perform detection of tones within files? I.e. I can generate a WAV file via CLI that includes a tone, but what do I need to do to "scan" the file programmatically for embeds?

**Reply by ggerganov** [reply]
Thanks for the feedback!

There isn't such a tool currently, but it is easy to implement one.
Thanks for the idea - I will probably add it soon to the project.

**Reply by ggerganov** [reply]
Just added the [ggwave-from-file](https://github.com/ggerganov/ggwave/tree/master/examples/ggwave-from-file) example - it demonstrates how to decode messages from a WAV file.

**Comment by coachaymen** [comment]
Hi, I'd like to express my gratitude for the effort you've put in. I wanted to bring to your attention that while using Pico 2040, I encountered an issue with the Ultrasound protocol. Specifically, I was unable to utilize a sample rate of 48khz and a buffer size of 1024, despite testing it with Mic MAX9814 and Arduino nano rp2040 and Pico rp2040. Only a lower sample rate seems to be functional, as decoding at higher rates is taking too long, around 33-35 ms.

**Reply by ggerganov** [reply]
Hi, thanks for the feedback!
Yes, the 48kHz sample rate is somewhere near the limit of the RP2040.

I think at some point I was able to make it work, but I have lost the details.
You can try the following:

- Use Pico RP2040 with the [stock example](https://github.com/ggerganov/ggwave/blob/master/examples/rp2040-rx/rp2040-rx.ino)
- Disable all [Rx protocols](https://github.com/ggerganov/ggwave/blob/master/examples/rp2040-rx/rp2040-rx.ino#L152-L160)
- Enable only [GGWAVE_PROTOCOL_ULTRASOUND_FASTEST](https://github.com/ggerganov/ggwave/blob/master/examples/rp2040-rx/rp2040-rx.ino#L150)

The decoding speed depends on the number of enabled protocols, so the idea is to disable all unused protocols and leave on the ultrasound one. Try to see if this reduces the processing time. It has to be below `21ms` in order to work.

**Comment by diegohce** [comment]
@ggerganov Hi! 
I just pushed the repo https://github.com/diegohce/gogwave
gogwave are the go bindings (from the C API) to use ggwave.

let me know what you think.

Cheers!

**Comment by valooost** [comment]
Hey!
Nice Work!
I just wanted to ask if you have the waver as .exe too (Windows)
(I have it on IOS too, but i'm currently searching the Windows version)
Keep going and good luck for the future!
